### PR TITLE
feat(base): clearly indicate referenced documents

### PR DESCRIPTION
**Pull request by ninaandal**
### Description
Indicators that a document is a referenced document is added. Any document that has a reference to it will have a blue header/footer, a reference icon next to the title and a dialog that with information when publishing changes to a referenced document. 

A popover will be shown once for every user for every browser. Once the user dismiss it, it will not be shown again. 

These changes were necessary to inform the user of the consequences of making changes to a referenced document. 
Link to task: https://app.shortcut.com/sanity-io/story/21975/indicate-a-referenced-document-in-document-pane-list-view 

Screenshot 2022-08-01 at 12 50 24

### What to review

Go any file that is referenced in the studio to see the new features to indicate that the document is a referenced document. 
Check that the one time popover works. Press the publish button to see the dialog with a warning when publishing changes to a referenced document. 

### Notes for release

feat(base): clearly indicate referenced documents

### CODE REVIEW

File: packages/@sanity/desk-tool/src/panes/document/useDocumentIsReferenced.tsx [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import {createHookFromObservableFactory, getPublishedId} from '@sanity/base/_internal'
+import documentStore from 'part:@sanity/base/datastore/document'
+import {Observable} from 'rxjs'
+
+export const useDocumentsIsReferenced = createHookFromObservableFactory((documentId: string) => {
+  const totalClause = 'count(*[references($documentId)]) > 0'
```
Comment (sjelfull) [inline packages/@sanity/desk-tool/src/panes/document/useDocumentIsReferenced.tsx:6]: Will this work for referencing documents that you don't have permissions for?

Comment (ninaandal) [reply packages/@sanity/desk-tool/src/panes/document/useDocumentIsReferenced.tsx:6]: I don't know.. @rexxars, do you know?

File: packages/@sanity/desk-tool/src/panes/document/documentPanel/header/DocumentPanelHeader.tsx [hunk 2]
```diff
@@ -18,118 +18,125 @@ import {DocumentHeaderTitle} from './DocumentHeaderTitle'
...
-      views,
-    } = useDocumentPane()
-    const {revTime: rev} = historyController
-    const {features} = useDeskTool()
-    const {index, BackLink, hasGroupSiblings} = usePaneRouter()
-    const contextMenuItems = useMemo(() => menuItems.filter(isMenuButton), [menuItems])
-    const [isValidationOpen, setValidationOpen] = React.useState<boolean>(false)
-    const showTabs = views.length > 1
-    const showVersionMenu = features.reviewChanges
+  forwardRef(
```
Comment (vicbergquist) [inline packages/@sanity/desk-tool/src/panes/document/documentPanel/header/DocumentPanelHeader.tsx:28]: cant really see what changed below because the linting also changed. Is it possible to revert the changes not related to this PR?

File: packages/@sanity/desk-tool/src/panes/document/useDocumentIsReferenced.tsx (hunk 1 above)
Comment (rexxars) [reply packages/@sanity/desk-tool/src/panes/document/useDocumentIsReferenced.tsx:6]: Good point - it does not, unfortunately. I've escalated this internally, we'll discuss further there.

File: packages/@sanity/desk-tool/src/panes/document/useDocumentIsReferenced.tsx [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import {createHookFromObservableFactory, getPublishedId} from '@sanity/base/_internal'
+import documentStore from 'part:@sanity/base/datastore/document'
+import {Observable} from 'rxjs'
+
+export const useDocumentsIsReferenced = createHookFromObservableFactory((documentId: string) => {
```
Comment (rexxars) [inline packages/@sanity/desk-tool/src/panes/document/useDocumentIsReferenced.tsx:5]: I was a bit hasty in suggesting we create a separate hook for this - it looks like we probably want to reuse the [useReferringDocuments](https://github.com/sanity-io/sanity/blob/next/packages/@sanity/desk-tool/src/components/confirmDeleteDialog/useReferringDocuments.ts#L140) hook, since it also handles cross-project references.

Ideally we'd lift the hook out of `confirmDeleteDialog` since it would then be used in multiple locations.

File: packages/@sanity/desk-tool/src/panes/document/statusBar/dialogs/ConfirmDialog.tsx [hunk 4]
```diff
@@ -19,15 +20,17 @@ export function ConfirmDialog(props: {
   const {dialog, referenceElement} = props
 
   return (
```
Comment (rexxars) [inline packages/@sanity/desk-tool/src/panes/document/statusBar/dialogs/ConfirmDialog.tsx:22]: Maybe add a commenting explaining the theme color provider usage?

```suggestion
  // Wrapping in ThemeColorProvider to avoid the ConfirmDialog inheriting tone of the footer
  return (
```

File: packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    setHidePopover(true)
+  }
+
+  return (
+    <ThemeColorProvider tone="default">
+      <Box marginLeft={2}>
+        <Popover
+          content={<InnerPopover onClose={onClose} />}
+          placement="bottom-start"
+          referenceElement={document.getElementById('referenceTitle')}
```
Comment (rexxars) [inline packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx:36]: This should use a reference (`useRef`) - multiple document panes can be open at the same time, which will create multiple elements with the same ID

File: packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+          portal
+          open
+        >
+          <Box display="flex">
+            <TitleText id="referenceTitle" tabIndex={0} textOverflow="ellipsis" weight="semibold">
```
Comment (rexxars) [inline packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx]: See above comment

File: packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+          </Box>
+        </Popover>
+      </Box>
+    </ThemeColorProvider>
+  )
+}
+
+function InnerPopover({onClose}: {onClose?: () => void}) {
+  return (
+    <StyledBox as={'div'}>
```
Comment (rexxars) [inline packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx:56]: Stylistically, we prefer that static string values use the regular XML-like notation, even if it has no impact on the actual output.

```suggestion
    <StyledBox as="div">
```

File: packages/@sanity/desk-tool/src/panes/document/DocumentPane.tsx [hunk 8]
```diff
@@ -67,7 +68,8 @@ export const DocumentPane = memo(function DocumentPane(props: DocumentPaneProvid
   const [templatePermissions, isTemplatePermissionsLoading] = useUnstableTemplatePermissions(
     getNewDocumentOptions()
   )
-  const isLoaded = isDocumentLoaded && !isTemplatePermissionsLoading
+  const [isReferenced, isReferenceLoading] = useDocumentsIsReferenced(options.id)
```
Comment (rexxars) [inline packages/@sanity/desk-tool/src/panes/document/DocumentPane.tsx:71]: Seeing the amount of places this needs to be ["prop-drilled"](https://blog.logrocket.com/solving-prop-drilling-react-apps), I am wondering if perhaps it would make sense to lift this up to the [DocumentPaneProvider](https://github.com/sanity-io/sanity/blob/next/packages/@sanity/desk-tool/src/panes/document/DocumentPaneProvider.tsx) and attach it to the [DocumentPaneContextValue](https://github.com/sanity-io/sanity/blob/next/packages/@sanity/desk-tool/src/panes/document/DocumentPaneContext.ts#L13). That way, you could utilize the `useDocumentPane()` hook in order to get the information instead of relying on the prop being passed down everywhere you might need it.

File: packages/@sanity/desk-tool/src/panes/document/documentPanel/DocumentPanel.tsx [hunk 9]
```diff
@@ -63,6 +64,7 @@ export const DocumentPanel = function DocumentPanel(props: DocumentPanelProps) {
   const headerRect = useElementRect(headerElement)
   const portalRef = useRef<HTMLDivElement | null>(null)
   const [documentScrollElement, setDocumentScrollElement] = useState<HTMLDivElement | null>(null)
+  const [isReferencedDocument] = useDocumentsIsReferenced(documentId)
```
Comment (rexxars) [inline packages/@sanity/desk-tool/src/panes/document/documentPanel/DocumentPanel.tsx:67]: I believe this will end up creating two listeners for the same information - I would prefer that we either receive the `isReferencedDocument` through a prop, or as stated above, through the `DocumentPaneContext`

File: packages/@sanity/desk-tool/src/panes/document/documentPanel/header/DocumentPanelHeader.tsx (hunk 2 above)
Comment (rexxars) [reply packages/@sanity/desk-tool/src/panes/document/documentPanel/header/DocumentPanelHeader.tsx:28]: Prettier is "at fault" here - not much to do about it, I think. One alternative would be to refactor the code a bit so prettier doesn't wrap it further, heh

File: packages/@sanity/desk-tool/src/panes/document/documentPanel/header/DocumentPanelHeader.tsx [hunk 10]
```diff
@@ -18,118 +18,125 @@ import {DocumentHeaderTitle} from './DocumentHeaderTitle'
...
+    (
+      {rootElement, isReferencedDocument}: DocumentPanelHeaderProps,
+      ref: React.ForwardedRef<HTMLDivElement>
+    ) => {
```
Comment (rexxars) [inline packages/@sanity/desk-tool/src/panes/document/documentPanel/header/DocumentPanelHeader.tsx:32]: This is very nitpicky and only addressing Victorias note above - but theoretically I think you might force prettier to keep formatting if you do something like this:

```suggestion
  forwardRef((headerProps: DocumentPanelHeaderProps, ref: React.ForwardedRef<HTMLDivElement>) => {
  const {rootElement, isReferencedDocument} = headerProps
```

**Review by rexxars — CHANGES_REQUESTED** [review]

File: packages/@sanity/desk-tool/src/panes/document/DocumentPane.tsx (hunk 8 above)
Comment (ninaandal) [reply packages/@sanity/desk-tool/src/panes/document/DocumentPane.tsx:71]: This makes SO much sense - thanks for pointing this out 🙌

File: packages/@sanity/desk-tool/src/panes/document/useDocumentIsReferenced.tsx (hunk 3 above)
Comment (ninaandal) [reply packages/@sanity/desk-tool/src/panes/document/useDocumentIsReferenced.tsx:5]: Makes sense - I'll delete the other hook and use `useReferringDocuments` instead 🙌

File: packages/@sanity/desk-tool/src/actions/PublishAction.tsx [hunk 11]
```diff
@@ -152,6 +189,15 @@ export const PublishAction: DocumentActionComponent = (props) => {
...
       : title,
     shortcut: disabled || publishScheduled ? null : 'Ctrl+Alt+P',
-    onHandle: handle,
+    onHandle: documentIsReferenced ? () => setShowConfirmPublishReferencedDocument(true) : handle,
+    dialog: showConfirmPublishReferencedDocument ? dialog : undefined,
   }
 }
+
+const LinkBox = styled.div`
+  border-color: #e8f1fe;
```
Comment (vicbergquist) [inline packages/@sanity/desk-tool/src/actions/PublishAction.tsx:198]: I think this should use a color from the theme values instead. Same with the grey below for the border 👍

File: packages/@sanity/desk-tool/src/components/pane/PaneHeader.styles.tsx [hunk 12]
```diff
@@ -79,3 +79,18 @@ export const TabsBox = styled(Box)(({theme}: {theme: Theme}) => {
     }
   `
 })
+
+export const LinkCircle = styled.div`
+  border: 1px solid grey;
```
Comment (vicbergquist) [inline packages/@sanity/desk-tool/src/components/pane/PaneHeader.styles.tsx:84]: Same here, should use theme colors!

File: packages/@sanity/desk-tool/src/components/pane/PaneHeader.styles.tsx [hunk 13]
```diff
@@ -79,3 +79,18 @@ export const TabsBox = styled(Box)(({theme}: {theme: Theme}) => {
...
+  border-radius: 50%;
+  width: 24px;
+  height: 24px;
+`
+
+export const HorizontalLine = styled.div`
+  border-top: 1px solid #ced2d9;
```
Comment (vicbergquist) [inline packages/@sanity/desk-tool/src/components/pane/PaneHeader.styles.tsx:91]: should use theme color

File: packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+            </LinkCircle>
+          </Box>
+          <Heading size={1}>This is a referenced document</Heading>
+        </Flex>
+        <Text>Changes to this type of document may affect other documents that reference it</Text>
+      </Box>
+      <HorizontalLine />
+      <Box as="li" padding={3}>
+        <Stack space={2}>
+          <Button onClick={onClose} tone="primary" text="Got it" />
```
Comment (vicbergquist) [inline packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx:71]: It's currently not possible to close the popover with keyboard only, would be great if this button gets focus when it's rendered so that it becomes keyboard accessible

File: packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+            <LinkCircle>
+              <LinkIcon fontSize={'24'} />
+            </LinkCircle>
+          </Box>
+          <Heading size={1}>This is a referenced document</Heading>
+        </Flex>
+        <Text>Changes to this type of document may affect other documents that reference it</Text>
+      </Box>
+      <HorizontalLine />
+      <Box as="li" padding={3}>
```
Comment (vicbergquist) [inline packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx:69]: Any reason this is an `li` element? If yes, where is the parent `ul` or `ol`? Should be added if it's on purpose 👍

**Review by vicbergquist — CHANGES_REQUESTED** [review]
Great job! I have a couple of comments, mostly about using theme values for colors instead of hard-coded ones.

 Let me know if you have questions!

File: packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx (hunk 14 above)
Comment (ninaandal) [reply packages/@sanity/desk-tool/src/components/paneItem/ReferencedDocHeading.tsx:83]: I'll add the `autoFocus` prop to it 😊

**Review by vicbergquist — APPROVED** [review]

**Review by rexxars — APPROVED** [review]
👍
